node/src/components/consensus/highway_core/state.rs

```diff
@@ -361,54 +363,47 @@ impl<C: Context> State<C> {
     pub(crate) fn validate_vote(&self, swvote: &SignedWireVote<C>) -> Result<(), VoteError> {
         let wvote = &swvote.wire_vote;
         let creator = wvote.creator;
-        wvote.panorama.validate(self)?;
-        let mut justifications = wvote.panorama.iter_correct();
-        if !justifications.all(|vh| self.vote(vh).timestamp <= wvote.timestamp) {
+        let panorama = &wvote.panorama;
+        let timestamp = wvote.timestamp;
+        panorama.validate(self)?;
+        if panorama.iter_correct(self).any(|v| v.timestamp > timestamp) {
             return Err(VoteError::Timestamps);
         }
-        match wvote.panorama.get(creator) {
-            Observation::Faulty => {
-                error!("Vote from faulty validator should be rejected in `pre_validate_vote`.");
-                return Err(VoteError::FaultyCreator); // Should be unreachable.
+        if wvote.seq_number != panorama.next_seq_num(self, creator) {
+            return Err(VoteError::SequenceNumber);
+        }
+        let opt_prev_vote = panorama[creator].correct().map(|vh| self.vote(vh));
+        let prev_round_exp = opt_prev_vote.map_or(self.params.init_round_exp(), |v| v.round_exp);
+        let max_re = prev_round_exp.max(wvote.round_exp);
+        // The round exponent must only change one step at a time.
```

Comment:
I understand that we have made the decisions that round exp can change by only one at a time but why is that? If we have the allowed spread and minimum value why not allow for changing within those boundaries?

Reply:
That's a good question; it's indeed a bit arbitrary. You could even argue that we don't need either of those checks if we have a minimum round exponent that's not too small anyway.
I do think it's good to support configurations with very small minimal round exponents, but you're right that we could probably remove the other check now. It's just that the mechanism in the paper currently only allows steps of 1.

In the coming weeks we'll bring paper and code more in sync; let's see what comes out of that conversation. We can always loosen the validity conditions again.

Reply:
I was wondering whether allowing to move the round exponent within the boundaries of the allowable spread (around the median) wouldn't result in round exponents varying a lot between rounds. I wonder what would be the practical implication of that.

Reply:
Yes… I don't think it would actually be dangerous, but it certainly feels safer to leave the restriction in place for now.